## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [261]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0    1,096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [262]:
# View the column names in the data

df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [263]:
# Use the describe function to gather some basic statistics
df.describe()



client_id     order_id  order_week  order_year  unit_price  unit_cost  \
count  54,639.00    54,639.00   54,639.00   54,639.00   54,639.00  54,639.00   
mean   54,837.87 5,470,190.03       11.36    2,022.99      136.27      99.45   
std    25,487.44 2,599,806.93        7.02        0.08      183.87     133.16   
min    10,033.00 1,000,886.00        1.00    2,022.00        0.01       0.01   
25%    33,593.00 3,196,372.00        6.00    2,023.00       20.80      14.84   
50%    53,305.00 5,496,966.00       11.00    2,023.00       68.31      49.89   
75%    78,498.00 7,733,869.00       17.00    2,023.00      173.16     125.57   
max    99,984.00 9,998,480.00       52.00    2,023.00    1,396.23     846.27   

       unit_weight          qty  line_number  
count    54,639.00    54,639.00    54,639.00  
mean          5.00       570.26         2.98  
std           5.33    18,795.52         2.44  
min           0.00         0.00         0.00  
25%           1.45        32.00         1.00  
50%           3.24        68.00         3.00  
75%           6.89       170.00         5.00  
max          46.43 3,958,244.00         9.00

In [264]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.tail()


first       last                      job               phone  \
54634   Lauren      Reese           Radio producer  977.877.5272x11382   
54635  Derrick      Moore                 Musician        358.661.5483   
54636   Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637    Wanda    Solomon             Toxicologist       (311)767-4924   
54638   Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                    email  client_id  order_id  order_date  \
54634  late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635        derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636           gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637           solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                    belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  
54634            8  
54635            0  
54636            6  
54637            0  
54638            5

In [265]:
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [266]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [267]:
# For the category with the most entries,
# which subcategory had the most entries?
df= df.loc[df["category"] == "consumables"]

df['subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [273]:
# Which five clients had the most entries in the data?

df["client_id"].value_counts().head()



client_id
66037    95
24741    95
37715    95
49687    95
43607    94
Name: count, dtype: int64

In [274]:
# Store the client ids of those top 5 clients in a list.


top_5_clients = df["client_id"].value_counts().head().index.tolist()

print(top_5_clients)


[66037, 24741, 37715, 49687, 43607]


In [275]:
# How many total units (the qty column) did the
# client with the most entries order order?
df.groupby('client_id')['qty'].value_counts().sum()



23538

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [279]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty


line_subtotal = df['unit_price']*df['qty']
df["line_subtotal"] = line_subtotal



df[["unit_price","qty", "line_subtotal"]].head()



unit_price  qty  line_subtotal
1       24.95   21         523.95
3       36.42   29       1,056.18
4      195.10   20       3,902.00
6      245.56   56      13,751.36
7       20.99   22         461.78

In [280]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
total_weight = df['unit_weight']*df['qty']
df["total_weight"] = total_weight
def shipping(total_weight):
    if total_weight < 50:
        return total_weight * 10
    return total_weight * 7

df['Shipping_price'] = df['total_weight'].apply(shipping)
df[["unit_price","qty","unit_weight", "total_weight","Shipping_price"]].head()

unit_price  qty  unit_weight  total_weight  Shipping_price
1       24.95   21         1.49         31.29          312.90
3       36.42   29         1.23         35.67          356.70
4      195.10   20        46.43        928.60        6,500.20
6      245.56   56         1.73         96.88          678.16
7       20.99   22         5.12        112.64          788.48

In [281]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

line_price = (df['Shipping_price'] + df['line_subtotal']) * 1.0925
df["line_price"] = line_price

df.head()

first      last                    job                phone  \
1    Tiffany     Myers        Music therapist     201.442.4543x942   
3     Nathan     Baker  Accounting technician     827-788-8123x012   
4  Christina  Schwartz           Chiropractor         265-829-3643   
6      Brian     Moore                Barista  +1-893-400-2733x982   
7      Jesus    Wright     Theme park manager   605.947.9467x66513   

                                 email  client_id  order_id  order_date  \
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
6          moore.official@martinez.com      87213   8080280  2023-04-19   
7        wright.jesus.7579@meadows.com      46709   8946782  2023-02-14   

   order_week  order_year  ...       subcategory unit_price unit_cost  \
1          20        2023  ...              pens      24.95     15.09   
3          17        2023  ...              pens      36.42     24.85   
4          21        2023  ...              misc     195.10    108.17   
6          16        2023  ...  kitchen supplies     245.56    231.95   
7           7        2023  ...              pens      20.99     16.97   

   unit_weight  qty  line_number  line_subtotal  total_weight  Shipping_price  \
1         1.49   21            0         523.95         31.29          312.90   
3         1.23   29            3       1,056.18         35.67          356.70   
4        46.43   20            1       3,902.00        928.60        6,500.20   
6         1.73   56            5      13,751.36         96.88          678.16   
7         5.12   22            4         461.78        112.64          788.48   

   line_price  
1      914.26  
3    1,543.57  
4   11,364.40  
6   15,764.25  
7    1,365.91  

[5 rows x 22 columns]

In [282]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

line_cost = df['unit_cost'] * df['qty'] + df['Shipping_price'] 
df["line_cost"] = line_cost

df.head()

first      last                    job                phone  \
1    Tiffany     Myers        Music therapist     201.442.4543x942   
3     Nathan     Baker  Accounting technician     827-788-8123x012   
4  Christina  Schwartz           Chiropractor         265-829-3643   
6      Brian     Moore                Barista  +1-893-400-2733x982   
7      Jesus    Wright     Theme park manager   605.947.9467x66513   

                                 email  client_id  order_id  order_date  \
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
6          moore.official@martinez.com      87213   8080280  2023-04-19   
7        wright.jesus.7579@meadows.com      46709   8946782  2023-02-14   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
1          20        2023  ...      24.95     15.09        1.49   21   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   
6          16        2023  ...     245.56    231.95        1.73   56   
7           7        2023  ...      20.99     16.97        5.12   22   

   line_number  line_subtotal  total_weight  Shipping_price  line_price  \
1            0         523.95         31.29          312.90      914.26   
3            3       1,056.18         35.67          356.70    1,543.57   
4            1       3,902.00        928.60        6,500.20   11,364.40   
6            5      13,751.36         96.88          678.16   15,764.25   
7            4         461.78        112.64          788.48    1,365.91   

   line_cost  
1     629.79  
3   1,077.35  
4   8,663.60  
6  13,667.36  
7   1,161.82  

[5 rows x 23 columns]

In [283]:
# Create a column for the profit of
# each line using line cost and line price

line_profit =  df['line_price'] - df['line_cost']
df["line_profit"] = line_profit

df.head()


first      last                    job                phone  \
1    Tiffany     Myers        Music therapist     201.442.4543x942   
3     Nathan     Baker  Accounting technician     827-788-8123x012   
4  Christina  Schwartz           Chiropractor         265-829-3643   
6      Brian     Moore                Barista  +1-893-400-2733x982   
7      Jesus    Wright     Theme park manager   605.947.9467x66513   

                                 email  client_id  order_id  order_date  \
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
6          moore.official@martinez.com      87213   8080280  2023-04-19   
7        wright.jesus.7579@meadows.com      46709   8946782  2023-02-14   

   order_week  order_year  ... unit_cost unit_weight qty  line_number  \
1          20        2023  ...     15.09        1.49  21            0   
3          17        2023  ...     24.85        1.23  29            3   
4          21        2023  ...    108.17       46.43  20            1   
6          16        2023  ...    231.95        1.73  56            5   
7           7        2023  ...     16.97        5.12  22            4   

   line_subtotal  total_weight  Shipping_price  line_price  line_cost  \
1         523.95         31.29          312.90      914.26     629.79   
3       1,056.18         35.67          356.70    1,543.57   1,077.35   
4       3,902.00        928.60        6,500.20   11,364.40   8,663.60   
6      13,751.36         96.88          678.16   15,764.25  13,667.36   
7         461.78        112.64          788.48    1,365.91   1,161.82   

   line_profit  
1       284.47  
3       466.22  
4     2,700.80  
6     2,096.89  
7       204.09  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [298]:
# Check your work using the totals above



## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [284]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

filtered_df = df[df['client_id'].apply(lambda x: x in top_5_clients)]

#filtered_df[["client_id", "unit_price","qty", "line_subtotal"]].head()

filtered_df

first     last                         job                  phone  \
52       Jared   Taylor               Haematologist           729-249-1032   
145      Jared   Taylor               Haematologist           729-249-1032   
216     Angela  Everett                   Osteopath       729-620-6192x875   
258    Heather      Day  Regulatory affairs officer      (251)730-3328x301   
321      Norma    Lopez          Arts administrator  001-935-686-5976x1888   
...        ...      ...                         ...                    ...   
54345    Jared   Taylor               Haematologist           729-249-1032   
54361   Angela  Everett                   Osteopath       729-620-6192x875   
54467    Norma    Lopez          Arts administrator  001-935-686-5976x1888   
54510   Kendra  Garrett      Pension scheme manager           808.257.2029   
54590    Jared   Taylor               Haematologist           729-249-1032   

                                 email  client_id  order_id  order_date  \
52                  taylorj@romero.com      37715   6174735  2023-01-07   
145                 taylorj@romero.com      37715   8442621  2023-02-01   
216             aeverett@rodriguez.com      66037   1203693  2023-04-09   
258                    dayh@hughes.net      43607   3598615  2023-03-29   
321       lopez.norma.2726@garcia.info      49687   1097098  2023-05-28   
...                                ...        ...       ...         ...   
54345               taylorj@romero.com      37715   3315044  2023-05-30   
54361           aeverett@rodriguez.com      66037   7733869  2023-01-23   
54467     lopez.norma.2726@garcia.info      49687   6641622  2023-01-21   
54510  garrettkendra8479@jefferson.com      24741   2994571  2023-03-03   
54590               taylorj@romero.com      37715   8442621  2023-02-01   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
52              1        2023  ...      2.39       10.49   116            5   
145             5        2023  ...     23.72        2.54    58            5   
216            14        2023  ...    166.09        0.26   157            2   
258            13        2023  ...     54.70        0.89    55            5   
321            21        2023  ...     36.96       10.69    23            3   
...           ...         ...  ...       ...         ...   ...          ...   
54345          22        2023  ...     14.52       12.42    87            0   
54361           4        2023  ...      9.29        5.67    23            6   
54467           3        2023  ...    222.48        0.19  1054            4   
54510           9        2023  ...     78.22        1.64   173            0   
54590           5        2023  ...     28.34        0.54    34            6   

       line_subtotal  total_weight  Shipping_price  line_price  line_cost  \
52            541.72      1,216.84        8,517.88    9,897.61   8,795.12   
145         1,530.62        147.32        1,031.24    2,798.83   2,407.00   
216        47,393.59         40.82          408.20   52,223.46  26,484.33   
258         3,961.10         48.95          489.50    4,862.28   3,498.00   
321         1,151.38        245.87        1,721.09    3,138.17   2,571.17   
...              ...           ...             ...         ...        ...   
54345       1,591.23      1,080.54        7,563.78   10,001.85   8,827.02   
54361         316.02        130.41          912.87    1,342.56   1,126.54   
54467     261,381.46        200.26        1,401.82  287,090.73 235,895.74   
54510      15,006.02        283.72        1,986.04   18,563.83  15,518.10   
54590       1,111.46         18.36          183.60    1,414.85   1,147.16   

       line_profit  
52        1,102.49  
145         391.83  
216      25,739.13  
258       1,364.28  
321         567.00  
...            ...  
54345     1,174.83  
54361       216.02  
54467    51,194.99  
54510     3,045.73  
54590       267.69  

[474 rows x 24 columns]

In [285]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
total_df = filtered_df.groupby('client_id').agg({
    'qty': 'sum',
    'Shipping_price': 'sum',
    'line_cost': 'sum',
    'line_profit': 'sum'
})
summary_df = total_df.sort_values(by='line_profit', ascending=False)

summary_df.head()

qty  Shipping_price     line_cost  line_profit
client_id                                                   
49687      153026    2,529,273.89 13,711,326.45 7,640,445.22
66037       18817      801,781.18  2,817,914.86 1,466,826.00
24741       28525    1,079,120.99  3,044,062.24 1,158,183.00
43607       17884      742,918.25  3,077,457.05   948,289.45
37715       12586    1,170,963.28  2,520,723.30   911,553.83

In [307]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

client_df = summary_df.rename(columns={"client_id":"Client ID",
                                        "qty": "Quantity",
                                        "shipping_price": "Shipping",
                                        "line_price": "Total Price",
                                        "line_cost": "Total Revenue",
                                        "line_profit": "Total Profit", })

def scale_to_millions(column):

    return '{:,.2f}M'.format('Total Price' /1000000)
#client_df['Total Price'] = client_df.apply(scale_to_millions, axis = 1)

client_df.head()

Quantity  Shipping_price  Total Revenue  Total Profit
client_id                                                       
49687        153026    2,529,273.89  13,711,326.45  7,640,445.22
66037         18817      801,781.18   2,817,914.86  1,466,826.00
24741         28525    1,079,120.99   3,044,062.24  1,158,183.00
43607         17884      742,918.25   3,077,457.05    948,289.45
37715         12586    1,170,963.28   2,520,723.30    911,553.83

In [299]:
# Sort the updated data by "Total Profit" form highest to lowest
client_df.sort_values(by='Total Profit', ascending=False)

Quantity  Shipping_price  Total Revenue  Total Profit
client_id                                                       
24741        239862    5,126,448.37  45,688,899.71 36,579,992.27
38378         73667    3,429,455.40   9,634,720.98  3,271,829.90
66037         43018    1,395,151.85   7,004,482.98  3,255,031.82
46820         75768    1,601,448.84   7,007,191.64  2,736,602.68
33615         64313    1,828,834.89   6,175,163.91  2,201,980.79